In [8]:
import pandas as pd

happiness_2015 = pd.read_csv("data/raw/2015.csv")

happiness_2015.head()

,Country,Region,Happiness Rank,Happiness Score,Standard Error,Economy (GDP per Capita),Family,Health (Life Expectancy),Freedom,Trust (Government Corruption),Generosity,Dystopia Residual
0,Switzerland,Western Europe,1,7.587,0.03411,1.39651,1.34951,0.94143,0.66557,0.41978,0.29678,2.51738
1,Iceland,Western Europe,2,7.561,0.04884,1.30232,1.40223,0.94784,0.62877,0.14145,0.43630,2.70201
2,Denmark,Western Europe,3,7.527,0.03328,1.32548,1.36058,0.87464,0.64938,0.48357,0.34139,2.49204
3,Norway,Western Europe,4,7.522,0.03880,1.45900,1.33095,0.88521,0.66973,0.36503,0.34699,2.46531
4,Canada,North America,5,7.427,0.03553,1.32629,1.32261,0.90563,0.63297,0.32957,0.45811,2.45176


In [9]:
happiness_2016 = pd.read_csv("data/raw/2016.csv")
happiness_2017 = pd.read_csv("data/raw/2017.csv")
happiness_2018 = pd.read_csv("data/raw/2018.csv")
happiness_2019 = pd.read_csv("data/raw/2019.csv")

In [10]:
for year, df in [
    (2015, happiness_2015),
    (2016, happiness_2016),
    (2017, happiness_2017),
    (2018, happiness_2018),
    (2019, happiness_2019)
]:
    print(year, df.shape)
    print(df.columns.tolist())
    print()

2015 (158, 12)
['Country', 'Region', 'Happiness Rank', 'Happiness Score', 'Standard Error', 'Economy (GDP per Capita)', 'Family', 'Health (Life Expectancy)', 'Freedom', 'Trust (Government Corruption)', 'Generosity', 'Dystopia Residual']

2016 (157, 13)
['Country', 'Region', 'Happiness Rank', 'Happiness Score', 'Lower Confidence Interval', 'Upper Confidence Interval', 'Economy (GDP per Capita)', 'Family', 'Health (Life Expectancy)', 'Freedom', 'Trust (Government Corruption)', 'Generosity', 'Dystopia Residual']

2017 (155, 12)
['Country', 'Happiness.Rank', 'Happiness.Score', 'Whisker.high', 'Whisker.low', 'Economy..GDP.per.Capita.', 'Family', 'Health..Life.Expectancy.', 'Freedom', 'Generosity', 'Trust..Government.Corruption.', 'Dystopia.Residual']

2018 (156, 9)
['Overall rank', 'Country or region', 'Score', 'GDP per capita', 'Social support', 'Healthy life expectancy', 'Freedom to make life choices', 'Generosity', 'Perceptions of corruption']

2019 (156, 9)
['Overall rank', 'Country or 

In [11]:
for year, df in [
    (2015, happiness_2015),
    (2016, happiness_2016),
    (2017, happiness_2017),
    (2018, happiness_2018),
    (2019, happiness_2019)
]:
    print(year)
    print("Duplicados:", df.duplicated().sum())
    print("Valores nulos:", df.isnull().sum().sum())
    print()

2015
Duplicados: 0
Valores nulos: 0

2016
Duplicados: 0
Valores nulos: 0

2017
Duplicados: 0
Valores nulos: 0

2018
Duplicados: 0
Valores nulos: 1

2019
Duplicados: 0
Valores nulos: 0



In [ ]:
# buscar el valor nulo
happiness_2018[happiness_2018.isnull().any(axis=1)]

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
19,20,United Arab Emirates,6.774,2.096,0.776,0.67,0.284,0.186,NaN


# Valores nulos

Se detectó un único valor nulo en los datos del World Happiness Report: la variable `Perceptions of corruption` de United Arab Emirates en 2018.

Se decide mantener el registro y conservar el valor como nulo, ya que el resto de la información del país está disponible y eliminar la fila supondría perder datos útiles para el análisis.

In [13]:
from cleaning import clean_happiness_2015_2016, clean_happiness_2017

happiness_2015_clean = clean_happiness_2015_2016(happiness_2015, 2015)
happiness_2016_clean = clean_happiness_2015_2016(happiness_2016, 2016)
happiness_2017_clean = clean_happiness_2017(happiness_2017, 2017)

happiness_2017_clean.head()

,country,happiness_rank,happiness_score,gdp_score,social_support,life_expectancy,freedom,generosity,corruption,year
0,Norway,1,7.537,1.616463,1.533524,0.796667,0.635423,0.362012,0.315964,2017
1,Denmark,2,7.522,1.482383,1.551122,0.792566,0.626007,0.355280,0.400770,2017
2,Iceland,3,7.504,1.480633,1.610574,0.833552,0.627163,0.475540,0.153527,2017
3,Switzerland,4,7.494,1.564980,1.516912,0.858131,0.620071,0.290549,0.367007,2017
4,Finland,5,7.469,1.443572,1.540247,0.809158,0.617951,0.245483,0.382612,2017


In [15]:
import cleaning
import importlib

importlib.reload(cleaning)

<module 'cleaning' from 'c:\\Users\\hp\\Desktop\\IronHackBeatriz\\Proyectos\\Proyecto_SQL\\SQL_Project\\cleaning.py'>

In [16]:
from cleaning import clean_happiness_2018_2019

happiness_2018_clean = clean_happiness_2018_2019(happiness_2018, 2018)
happiness_2019_clean = clean_happiness_2018_2019(happiness_2019, 2019)

happiness_2018_clean.head()

,country,happiness_rank,happiness_score,gdp_score,social_support,life_expectancy,freedom,generosity,corruption,year
0,Finland,1,7.632,1.305,1.592,0.874,0.681,0.202,0.393,2018
1,Norway,2,7.594,1.456,1.582,0.861,0.686,0.286,0.340,2018
2,Denmark,3,7.555,1.351,1.590,0.868,0.683,0.284,0.408,2018
3,Iceland,4,7.495,1.343,1.644,0.914,0.677,0.353,0.138,2018
4,Switzerland,5,7.487,1.420,1.549,0.927,0.660,0.256,0.357,2018


In [17]:
# Unimos los cinco años en un único DataFrame
happiness = pd.concat([
    happiness_2015_clean,
    happiness_2016_clean,
    happiness_2017_clean,
    happiness_2018_clean,
    happiness_2019_clean
], ignore_index=True)

# Comprobamos el tamaño del DataFrame final
happiness.shape

(782, 10)

In [18]:
# Comprobamos cuántos registros tenemos por año
happiness["year"].value_counts().sort_index()

year
2015    158
2016    157
2017    155
2018    156
2019    156
Name: count, dtype: int64

In [19]:
# Comprobamos el DataFrame final
happiness.info()

<class 'pandas.DataFrame'>
RangeIndex: 782 entries, 0 to 781
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country          782 non-null    str    
 1   happiness_rank   782 non-null    int64  
 2   happiness_score  782 non-null    float64
 3   gdp_score        782 non-null    float64
 4   social_support   782 non-null    float64
 5   life_expectancy  782 non-null    float64
 6   freedom          782 non-null    float64
 7   generosity       782 non-null    float64
 8   corruption       781 non-null    float64
 9   year             782 non-null    int64  
dtypes: float64(7), int64(2), str(1)
memory usage: 67.5 KB


# Fuente 2: API del Banco Mundial

La segunda fuente de datos utilizada en el proyecto es la API del Banco Mundial. Esta fuente se utilizará para obtener indicadores económicos y sociales de los países entre 2015 y 2019 y relacionarlos posteriormente con los datos del World Happiness Report.

In [20]:
# Importamos requests para realizar peticiones a la API
import requests

# URL de la API del Banco Mundial
url = "https://api.worldbank.org/v2/country/ESP/indicator/NY.GDP.PCAP.CD"

# Parámetros de la consulta
params = {
    "date": "2015:2019",
    "format": "json"
}

# Realizamos la petición
response = requests.get(url, params=params)

# Comprobamos si la petición ha funcionado
response.status_code

200

In [21]:
data = response.json()
data

[{'page': 1,
  'pages': 1,
  'per_page': 50,
  'total': 5,
  'sourceid': '2',
  'lastupdated': '2026-07-13'},
 [{'indicator': {'id': 'NY.GDP.PCAP.CD',
    'value': 'GDP per capita (current US$)'},
   'country': {'id': 'ES', 'value': 'Spain'},
   'countryiso3code': 'ESP',
   'date': '2019',
   'value': 29786.5245689007,
   'unit': '',
   'obs_status': '',
   'decimal': 1},
  {'indicator': {'id': 'NY.GDP.PCAP.CD',
    'value': 'GDP per capita (current US$)'},
   'country': {'id': 'ES', 'value': 'Spain'},
   'countryiso3code': 'ESP',
   'date': '2018',
   'value': 30602.4212881113,
   'unit': '',
   'obs_status': '',
   'decimal': 1},
  {'indicator': {'id': 'NY.GDP.PCAP.CD',
    'value': 'GDP per capita (current US$)'},
   'country': {'id': 'ES', 'value': 'Spain'},
   'countryiso3code': 'ESP',
   'date': '2017',
   'value': 28381.3425596906,
   'unit': '',
   'obs_status': '',
   'decimal': 1},
  {'indicator': {'id': 'NY.GDP.PCAP.CD',
    'value': 'GDP per capita (current US$)'},
   'coun

In [22]:
# Guardamos únicamente la parte de la respuesta que contiene los datos
gdp_data = data[1]

# Convertimos los datos de la API en un DataFrame
gdp_spain = pd.DataFrame(gdp_data)

# Mostramos el DataFrame
gdp_spain

,indicator,country,countryiso3code,date,value,unit,obs_status,decimal
0,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ES', 'value': 'Spain'}",ESP,2019,29786.524569,,,1
1,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ES', 'value': 'Spain'}",ESP,2018,30602.421288,,,1
2,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ES', 'value': 'Spain'}",ESP,2017,28381.342560,,,1
3,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ES', 'value': 'Spain'}",ESP,2016,26755.606115,,,1
4,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ES', 'value': 'Spain'}",ESP,2015,25982.441620,,,1


In [ ]:
# Extraemos el nombre del país del diccionario de la columna country

#En la columna country, recorre cada registro, extrae su value y guarda el resultado de nuevo en country

gdp_spain["country"] = gdp_spain["country"].apply(lambda x: x["value"])

# Comprobamos el resultado
gdp_spain

,indicator,country,countryiso3code,date,value,unit,obs_status,decimal
0,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...",Spain,ESP,2019,29786.524569,,,1
1,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...",Spain,ESP,2018,30602.421288,,,1
2,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...",Spain,ESP,2017,28381.342560,,,1
3,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...",Spain,ESP,2016,26755.606115,,,1
4,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...",Spain,ESP,2015,25982.441620,,,1


In [24]:
# Seleccionamos únicamente las columnas que necesitamos
gdp_spain = gdp_spain[[
    "country",
    "countryiso3code",
    "date",
    "value"
]]

# Renombramos las columnas para que tengan nombres más claros
gdp_spain = gdp_spain.rename(columns={
    "countryiso3code": "iso3_code",
    "date": "year",
    "value": "gdp_per_capita"
})

gdp_spain

,country,iso3_code,year,gdp_per_capita
0,Spain,ESP,2019,29786.524569
1,Spain,ESP,2018,30602.421288
2,Spain,ESP,2017,28381.342560
3,Spain,ESP,2016,26755.606115
4,Spain,ESP,2015,25982.441620


In [25]:
# URL para obtener el PIB per cápita de todos los países
url = "https://api.worldbank.org/v2/country/all/indicator/NY.GDP.PCAP.CD"

# Indicamos los años, el formato y aumentamos los resultados por página
params = {
    "date": "2015:2019",
    "format": "json",
    "per_page": 20000
}

# petición a la API
response = requests.get(url, params=params)

# Comprobación
response.status_code

200

In [26]:
world_gdp_data = response.json()

In [27]:
# Mostramos la información general de la respuesta de la API
world_gdp_data[0]

{'page': 1,
 'pages': 1,
 'per_page': 20000,
 'total': 1325,
 'sourceid': '2',
 'lastupdated': '2026-07-13'}

In [28]:
# Comprobamos cuántos registros contiene la parte de datos
len(world_gdp_data[1])

1325

In [29]:
# Mostramos los primeros cinco registros
world_gdp_data[1][:5]

[{'indicator': {'id': 'NY.GDP.PCAP.CD',
   'value': 'GDP per capita (current US$)'},
  'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
  'countryiso3code': 'AFE',
  'date': '2019',
  'value': 1508.0315245887,
  'unit': '',
  'obs_status': '',
  'decimal': 1},
 {'indicator': {'id': 'NY.GDP.PCAP.CD',
   'value': 'GDP per capita (current US$)'},
  'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
  'countryiso3code': 'AFE',
  'date': '2018',
  'value': 1553.61794421394,
  'unit': '',
  'obs_status': '',
  'decimal': 1},
 {'indicator': {'id': 'NY.GDP.PCAP.CD',
   'value': 'GDP per capita (current US$)'},
  'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
  'countryiso3code': 'AFE',
  'date': '2017',
  'value': 1529.92307995549,
  'unit': '',
  'obs_status': '',
  'decimal': 1},
 {'indicator': {'id': 'NY.GDP.PCAP.CD',
   'value': 'GDP per capita (current US$)'},
  'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
  'countryiso3c

In [30]:
# Guardamos únicamente la parte que contiene los datos
gdp_data = world_gdp_data[1]

# Convertimos los datos de la API en un DataFrame
world_gdp = pd.DataFrame(gdp_data)

# Mostramos las primeras filas
world_gdp.head()

,indicator,country,countryiso3code,date,value,unit,obs_status,decimal
0,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ZH', 'value': 'Africa Eastern and Sout...",AFE,2019,1508.031525,,,1
1,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ZH', 'value': 'Africa Eastern and Sout...",AFE,2018,1553.617944,,,1
2,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ZH', 'value': 'Africa Eastern and Sout...",AFE,2017,1529.923080,,,1
3,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ZH', 'value': 'Africa Eastern and Sout...",AFE,2016,1335.743165,,,1
4,"{'id': 'NY.GDP.PCAP.CD', 'value': 'GDP per cap...","{'id': 'ZH', 'value': 'Africa Eastern and Sout...",AFE,2015,1498.715971,,,1


In [31]:
# Comprobamos el tamaño del DataFrame
world_gdp.shape

(1325, 8)

In [32]:
# Recargamos el archivo cleaning.py para reconocer la nueva función
import cleaning
import importlib

importlib.reload(cleaning)

# Importamos la función de limpieza del Banco Mundial
from cleaning import clean_world_bank

In [ ]:
# Limpiamos los datos de PIB per cápita
world_gdp_clean = clean_world_bank(
    world_gdp,
    "gdp_per_capita"
)

world_gdp_clean.head()

,country,iso3_code,year,gdp_per_capita
0,Africa Eastern and Southern,AFE,2019,1508.031525
1,Africa Eastern and Southern,AFE,2018,1553.617944
2,Africa Eastern and Southern,AFE,2017,1529.923080
3,Africa Eastern and Southern,AFE,2016,1335.743165
4,Africa Eastern and Southern,AFE,2015,1498.715971


In [34]:
# URL para obtener información sobre los países del Banco Mundial
countries_url = "https://api.worldbank.org/v2/country"

# Indicamos que queremos los datos en formato JSON
# y aumentamos el número de resultados por página
countries_params = {
    "format": "json",
    "per_page": 400
}

# Realizamos la petición a la API
countries_response = requests.get(countries_url, params=countries_params)

# Comprobamos que la petición se ha realizado correctamente
countries_response.status_code

200

In [35]:
# Convertimos la respuesta a formato JSON
countries_data = countries_response.json()

# Vemos la información general de la respuesta
countries_data[0]

{'page': 1, 'pages': 1, 'per_page': '400', 'total': 295}

In [36]:
# Mostramos el primer registro para entender su estructura
countries_data[1][0]

{'id': 'ABW',
 'iso2Code': 'AW',
 'name': 'Aruba',
 'region': {'id': 'LCN',
  'iso2code': 'ZJ',
  'value': 'Latin America & Caribbean '},
 'adminregion': {'id': '', 'iso2code': '', 'value': ''},
 'incomeLevel': {'id': 'HIC', 'iso2code': 'XD', 'value': 'High income'},
 'lendingType': {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'},
 'capitalCity': 'Oranjestad',
 'longitude': '-70.0167',
 'latitude': '12.5167'}

In [37]:
# Convertimos los metadatos de países en un DataFrame
countries_df = pd.DataFrame(countries_data[1])

# Buscamos el registro de Africa Eastern and Southern
countries_df[countries_df["name"] == "Africa Eastern and Southern"]

,id,iso2Code,name,region,adminregion,incomeLevel,lendingType,capitalCity,longitude,latitude
1,AFE,ZH,Africa Eastern and Southern,"{'id': 'NA', 'iso2code': 'NA', 'value': 'Aggre...","{'id': '', 'iso2code': '', 'value': ''}","{'id': 'NA', 'iso2code': 'NA', 'value': 'Aggre...","{'id': '', 'iso2code': '', 'value': 'Aggregates'}",,,


In [38]:
# Comprobamos el tamaño del DataFrame de países y entidades
countries_df.shape

(295, 10)

In [39]:
# Recargamos cleaning.py para reconocer la nueva función
import cleaning
import importlib

importlib.reload(cleaning)

# Importamos la nueva función
from cleaning import clean_world_bank_countries

# Limpiamos los metadatos de países
countries_clean = clean_world_bank_countries(countries_df)

# Comprobamos 
countries_clean.head()

,iso3_code,country,region,income_level
0,ABW,Aruba,Latin America & Caribbean,High income
1,AFG,Afghanistan,"Middle East, North Africa, Afghanistan & Pakistan",Low income
2,AGO,Angola,Sub-Saharan Africa,Lower middle income
3,ALB,Albania,Europe & Central Asia,Upper middle income
4,AND,Andorra,Europe & Central Asia,High income


In [40]:
# Comprobamos el tamaño después de eliminar los agregados
countries_clean.shape

(217, 4)

In [41]:
# Comprobamos que ya no existen registros clasificados como Aggregates
countries_clean[countries_clean["region"] == "Aggregates"]

,iso3_code,country,region,income_level


In [42]:
# Filtramos los datos de PIB para quedarnos únicamente con países reales
world_gdp_clean = world_gdp_clean[
    world_gdp_clean["iso3_code"].isin(countries_clean["iso3_code"])
]

# Reiniciamos el índice después del filtrado
world_gdp_clean = world_gdp_clean.reset_index(drop=True)

# Comprobamos el resultado
world_gdp_clean.head()

,country,iso3_code,year,gdp_per_capita
0,Afghanistan,AFG,2019,496.602504
1,Afghanistan,AFG,2018,491.337221
2,Afghanistan,AFG,2017,525.469771
3,Afghanistan,AFG,2016,522.082216
4,Afghanistan,AFG,2015,565.569730


In [43]:
# Comprobamos el nuevo tamaño
world_gdp_clean.shape

(1085, 4)

In [44]:
# Comprobamos que el agregado que vimos anteriormente ya no existe
world_gdp_clean[
    world_gdp_clean["country"] == "Africa Eastern and Southern"
]

,country,iso3_code,year,gdp_per_capita


In [45]:
# Comprobamos si existen valores nulos en los datos de PIB per cápita
world_gdp_clean.isnull().sum()

country            0
iso3_code          0
year               0
gdp_per_capita    29
dtype: int64

In [46]:
# Mostramos los registros que no tienen un valor de PIB per cápita
world_gdp_clean[world_gdp_clean["gdp_per_capita"].isnull()]

,country,iso3_code,year,gdp_per_capita
135,British Virgin Islands,VGB,2019,NaN
136,British Virgin Islands,VGB,2018,NaN
137,British Virgin Islands,VGB,2017,NaN
138,British Virgin Islands,VGB,2016,NaN
139,British Virgin Islands,VGB,2015,NaN
305,Eritrea,ERI,2019,NaN
306,Eritrea,ERI,2018,NaN
307,Eritrea,ERI,2017,NaN
308,Eritrea,ERI,2016,NaN
309,Eritrea,ERI,2015,NaN


In [47]:
# Obtenemos los países únicos del World Happiness Report
happiness_countries = happiness["country"].drop_duplicates()

# Comprobamos cuántos países diferentes aparecen
len(happiness_countries)

170

In [ ]:
# Buscamos los países de Happiness cuyo nombre no coincide
# con ningún nombre de los países del Banco Mundial
unmatched_countries = happiness[
    ~happiness["country"].isin(countries_clean["country"])
]["country"].drop_duplicates()

unmatched_countries

22                    Venezuela
30               Czech Republic
37                       Taiwan
44                     Slovakia
46                  South Korea
63                       Russia
65                 North Cyprus
71                    Hong Kong
74                      Vietnam
75                       Turkey
76                   Kyrgyzstan
90            Somaliland region
92                    Macedonia
98                         Laos
100                   Swaziland
107     Palestinian Territories
109                        Iran
119            Congo (Kinshasa)
134                       Egypt
135                       Yemen
138         Congo (Brazzaville)
150                 Ivory Coast
155                       Syria
172                 Puerto Rico
233                     Somalia
254           Somaliland Region
347    Taiwan Province of China
385     Hong Kong S.A.R., China
507           Trinidad & Tobago
527             Northern Cyprus
745                      Gambia
Name: co

# Valores nulos en los datos del Banco Mundial

Tras la extracción de los datos de PIB per cápita se detectaron 29 valores nulos.

Estos valores se mantienen temporalmente como nulos, ya que representan datos no disponibles en la fuente original. No se sustituyen por cero ni por valores estimados para evitar introducir información artificial en el análisis.

Posteriormente se comprobará si estos registros corresponden a países presentes en el World Happiness Report antes de decidir su tratamiento final.

In [49]:
# Convertimos los países que no coinciden en una lista
unmatched_list = unmatched_countries.tolist()

# Comprobamos cuántos países no coinciden por nombre
len(unmatched_list)

31

In [50]:
# Mostramos todos los países que no coinciden
for country in unmatched_list:
    print(country)

Venezuela
Czech Republic
Taiwan
Slovakia
South Korea
Russia
North Cyprus
Hong Kong
Vietnam
Turkey
Kyrgyzstan
Somaliland region
Macedonia
Laos
Swaziland
Palestinian Territories
Iran
Congo (Kinshasa)
Egypt
Yemen
Congo (Brazzaville)
Ivory Coast
Syria
Puerto Rico
Somalia
Somaliland Region
Taiwan Province of China
Hong Kong S.A.R., China
Trinidad & Tobago
Northern Cyprus
Gambia


In [51]:
# Recargamos cleaning.py
import cleaning
import importlib

importlib.reload(cleaning)

# Importamos la función
from cleaning import standardize_happiness_countries

# Estandarizamos los nombres
happiness_standardized = standardize_happiness_countries(happiness)

In [52]:
# Comprobamos qué países siguen sin coincidir
unmatched_after_mapping = happiness_standardized[
    ~happiness_standardized["country"].isin(countries_clean["country"])
]["country"].drop_duplicates()

for country in unmatched_after_mapping:
    print(country)

Taiwan
North Cyprus
Somaliland region
Puerto Rico
Somalia
Somaliland Region
Taiwan Province of China
Northern Cyprus


In [53]:
# Buscamos en los países del Banco Mundial nombres relacionados
# con los casos que todavía no hemos podido relacionar
countries_clean[
    countries_clean["country"].str.contains(
        "Taiwan|Cyprus|Somali|Puerto",
        case=False,
        na=False
    )
]

,iso3_code,country,region,income_level
48,CYP,Cyprus,Europe & Central Asia,High income
157,PRI,Puerto Rico (US),Latin America & Caribbean,High income
175,SOM,"Somalia, Fed. Rep.",Sub-Saharan Africa,Low income


In [55]:
# Recargamos cleaning.py después de modificarlo
import cleaning
import importlib

importlib.reload(cleaning)

# Volvemos a importar la función actualizada
from cleaning import standardize_happiness_countries

# Aplicamos de nuevo la estandarización
happiness_standardized = standardize_happiness_countries(happiness)

In [56]:
# Comprobamos qué países siguen sin coincidir
unmatched_after_mapping = happiness_standardized[
    ~happiness_standardized["country"].isin(countries_clean["country"])
]["country"].drop_duplicates()

for country in unmatched_after_mapping:
    print(country)

Taiwan
North Cyprus
Somaliland region
Somaliland Region
Taiwan Province of China
Northern Cyprus


In [57]:
# Añadimos a Happiness el código ISO3, la región y el nivel de ingresos
# utilizando el nombre estandarizado del país como columna de unión
happiness_with_country_data = happiness_standardized.merge(
    countries_clean,
    on="country",
    how="left"
)

happiness_with_country_data.head()

,country,happiness_rank,happiness_score,gdp_score,social_support,life_expectancy,freedom,generosity,corruption,year,iso3_code,region,income_level
0,Switzerland,1,7.587,1.39651,1.34951,0.94143,0.66557,0.29678,0.41978,2015,CHE,Europe & Central Asia,High income
1,Iceland,2,7.561,1.30232,1.40223,0.94784,0.62877,0.43630,0.14145,2015,ISL,Europe & Central Asia,High income
2,Denmark,3,7.527,1.32548,1.36058,0.87464,0.64938,0.34139,0.48357,2015,DNK,Europe & Central Asia,High income
3,Norway,4,7.522,1.45900,1.33095,0.88521,0.66973,0.34699,0.36503,2015,NOR,Europe & Central Asia,High income
4,Canada,5,7.427,1.32629,1.32261,0.90563,0.63297,0.45811,0.32957,2015,CAN,North America,High income


In [58]:
# Comprobamos cuántas filas de Happiness no han recibido código ISO3
happiness_with_country_data["iso3_code"].isnull().sum()

np.int64(12)

In [59]:
# Mostramos las filas de Happiness que no han recibido código ISO3
happiness_without_iso = happiness_with_country_data[
    happiness_with_country_data["iso3_code"].isnull()
]

# Mostramos solo las columnas necesarias para revisar estos casos
happiness_without_iso[[
    "country",
    "year",
    "happiness_score"
]]

,country,year,happiness_score
37,Taiwan,2015,6.298
65,North Cyprus,2015,5.695
90,Somaliland region,2015,5.057
192,Taiwan,2016,6.379
219,North Cyprus,2016,5.771
254,Somaliland Region,2016,5.057
347,Taiwan Province of China,2017,6.422
375,North Cyprus,2017,5.810
495,Taiwan,2018,6.441
527,Northern Cyprus,2018,5.835


In [60]:
# Creamos un DataFrame con las filas de Happiness
# que sí hemos podido relacionar con un código ISO3
happiness_matched = happiness_with_country_data[
    happiness_with_country_data["iso3_code"].notnull()
].copy()

# Reiniciamos el índice
happiness_matched = happiness_matched.reset_index(drop=True)

# Comprobamos el tamaño
happiness_matched.shape

(770, 13)

In [61]:
# Unimos los datos de Happiness con los datos de PIB del Banco Mundial
# utilizando como claves el código ISO3 del país y el año
happiness_gdp = happiness_matched.merge(
    world_gdp_clean[["iso3_code", "year", "gdp_per_capita"]],
    on=["iso3_code", "year"],
    how="left"
)

happiness_gdp.head()

,country,happiness_rank,happiness_score,gdp_score,social_support,life_expectancy,freedom,generosity,corruption,year,iso3_code,region,income_level,gdp_per_capita
0,Switzerland,1,7.587,1.39651,1.34951,0.94143,0.66557,0.29678,0.41978,2015,CHE,Europe & Central Asia,High income,85265.062052
1,Iceland,2,7.561,1.30232,1.40223,0.94784,0.62877,0.43630,0.14145,2015,ISL,Europe & Central Asia,High income,53505.696408
2,Denmark,3,7.527,1.32548,1.36058,0.87464,0.64938,0.34139,0.48357,2015,DNK,Europe & Central Asia,High income,53094.013361
3,Norway,4,7.522,1.45900,1.33095,0.88521,0.66973,0.34699,0.36503,2015,NOR,Europe & Central Asia,High income,77220.952431
4,Canada,5,7.427,1.32629,1.32261,0.90563,0.63297,0.45811,0.32957,2015,CAN,North America,High income,43594.237908


In [62]:
# Comprobamos el tamaño después de unir las dos fuentes
happiness_gdp.shape

(770, 14)

In [63]:
# Comprobamos los valores nulos después de unir las dos fuentes
happiness_gdp.isnull().sum()

country            0
happiness_rank     0
happiness_score    0
gdp_score          0
social_support     0
life_expectancy    0
freedom            0
generosity         0
corruption         1
year               0
iso3_code          0
region             0
income_level       0
gdp_per_capita     5
dtype: int64

In [64]:
# Mostramos los registros de Happiness que no tienen
# un valor de PIB per cápita disponible en el Banco Mundial
happiness_gdp[
    happiness_gdp["gdp_per_capita"].isnull()
][[
    "country",
    "iso3_code",
    "year",
    "happiness_score",
    "gdp_per_capita"
]]

,country,iso3_code,year,happiness_score,gdp_per_capita
294,South Sudan,SSD,2016,3.832,NaN
453,South Sudan,SSD,2017,3.591,NaN
613,South Sudan,SSD,2018,3.254,NaN
764,"Yemen, Rep.",YEM,2019,3.380,NaN
769,South Sudan,SSD,2019,2.853,NaN


# Extracción del desempleo desde la API del Banco Mundial

Además del PIB per cápita, se obtienen datos sobre la tasa de desempleo para analizar su posible relación con los niveles de felicidad.

Primero se realiza una prueba con España para comprobar la respuesta de la API antes de realizar la extracción para todos los países.

In [65]:
# Indicador del Banco Mundial:
# desempleo total como porcentaje de la población activa
unemployment_url = "https://api.worldbank.org/v2/country/ESP/indicator/SL.UEM.TOTL.ZS"

# Indicamos los años que queremos y el formato de respuesta
unemployment_params = {
    "date": "2015:2019",
    "format": "json"
}

# Realizamos la petición a la API
unemployment_response = requests.get(
    unemployment_url,
    params=unemployment_params
)

# Comprobamos 
unemployment_response.status_code

200

In [66]:
# Convertimos la respuesta de la API a formato JSON
unemployment_data = unemployment_response.json()

# Mostramos los registros obtenidos
unemployment_data[1]

[{'indicator': {'id': 'SL.UEM.TOTL.ZS',
   'value': 'Unemployment, total (% of total labor force) (modeled ILO estimate)'},
  'country': {'id': 'ES', 'value': 'Spain'},
  'countryiso3code': 'ESP',
  'date': '2019',
  'value': 14.104,
  'unit': '',
  'obs_status': '',
  'decimal': 1},
 {'indicator': {'id': 'SL.UEM.TOTL.ZS',
   'value': 'Unemployment, total (% of total labor force) (modeled ILO estimate)'},
  'country': {'id': 'ES', 'value': 'Spain'},
  'countryiso3code': 'ESP',
  'date': '2018',
  'value': 15.255,
  'unit': '',
  'obs_status': '',
  'decimal': 1},
 {'indicator': {'id': 'SL.UEM.TOTL.ZS',
   'value': 'Unemployment, total (% of total labor force) (modeled ILO estimate)'},
  'country': {'id': 'ES', 'value': 'Spain'},
  'countryiso3code': 'ESP',
  'date': '2017',
  'value': 17.224,
  'unit': '',
  'obs_status': '',
  'decimal': 1},
 {'indicator': {'id': 'SL.UEM.TOTL.ZS',
   'value': 'Unemployment, total (% of total labor force) (modeled ILO estimate)'},
  'country': {'id': '

In [67]:
# Creamos la URL para obtener el desempleo de todos los países
world_unemployment_url = "https://api.worldbank.org/v2/country/all/indicator/SL.UEM.TOTL.ZS"

# Indicamos los años, el formato y un número de registros suficiente
world_unemployment_params = {
    "date": "2015:2019",
    "format": "json",
    "per_page": 20000
}

# Realizamos la petición a la API
world_unemployment_response = requests.get(
    world_unemployment_url,
    params=world_unemployment_params
)

# Comprobamos
world_unemployment_response.status_code

200

In [68]:
# Convertimos la respuesta de la API a formato JSON
world_unemployment_data = world_unemployment_response.json()

# Comprobamos 
world_unemployment_data[0]

{'page': 1,
 'pages': 1,
 'per_page': 20000,
 'total': 1325,
 'sourceid': '2',
 'lastupdated': '2026-07-13'}

In [69]:
# Extraemos los registros de desempleo
unemployment_records = world_unemployment_data[1]

# Convertimos los registros en un DataFrame
world_unemployment = pd.DataFrame(unemployment_records)

# Comprobamos tamaño
world_unemployment.shape

(1325, 8)

In [70]:
# Limpiamos los datos de desempleo utilizando
# la función reutilizable creada para los datos del Banco Mundial
world_unemployment_clean = clean_world_bank(
    world_unemployment,
    "unemployment_rate"
)

# Comprobamos 
world_unemployment_clean.head()

,country,iso3_code,year,unemployment_rate
0,Africa Eastern and Southern,AFE,2019,7.459106
1,Africa Eastern and Southern,AFE,2018,7.244843
2,Africa Eastern and Southern,AFE,2017,7.214915
3,Africa Eastern and Southern,AFE,2016,7.074429
4,Africa Eastern and Southern,AFE,2015,6.930941


In [71]:
# Filtramos los datos de desempleo para quedarnos
# únicamente con países reales
world_unemployment_clean = world_unemployment_clean[
    world_unemployment_clean["iso3_code"].isin(
        countries_clean["iso3_code"]
    )
].reset_index(drop=True)

# Comprobamos el tamaño después del filtrado
world_unemployment_clean.shape

(1085, 4)

In [72]:
# Comprobamos si existen valores nulos
# en la tasa de desempleo
world_unemployment_clean.isnull().sum()

country                0
iso3_code              0
year                   0
unemployment_rate    150
dtype: int64

In [73]:
# Unimos la tasa de desempleo con nuestro dataset de Happiness y PIB
# utilizando como claves el código ISO3 del país y el año
happiness_economic = happiness_gdp.merge(
    world_unemployment_clean[[
        "iso3_code",
        "year",
        "unemployment_rate"
    ]],
    on=["iso3_code", "year"],
    how="left"
)

happiness_economic.head()

,country,happiness_rank,happiness_score,gdp_score,social_support,life_expectancy,freedom,generosity,corruption,year,iso3_code,region,income_level,gdp_per_capita,unemployment_rate
0,Switzerland,1,7.587,1.39651,1.34951,0.94143,0.66557,0.29678,0.41978,2015,CHE,Europe & Central Asia,High income,85265.062052,4.801
1,Iceland,2,7.561,1.30232,1.40223,0.94784,0.62877,0.43630,0.14145,2015,ISL,Europe & Central Asia,High income,53505.696408,3.979
2,Denmark,3,7.527,1.32548,1.36058,0.87464,0.64938,0.34139,0.48357,2015,DNK,Europe & Central Asia,High income,53094.013361,6.278
3,Norway,4,7.522,1.45900,1.33095,0.88521,0.66973,0.34699,0.36503,2015,NOR,Europe & Central Asia,High income,77220.952431,4.296
4,Canada,5,7.427,1.32629,1.32261,0.90563,0.63297,0.45811,0.32957,2015,CAN,North America,High income,43594.237908,6.945


In [74]:
# Comprobamos que la unión no ha duplicado registros
happiness_economic.shape

(770, 15)

In [75]:
# Comprobamos los valores nulos después de añadir el desempleo
happiness_economic.isnull().sum()

country              0
happiness_rank       0
happiness_score      0
gdp_score            0
social_support       0
life_expectancy      0
freedom              0
generosity           0
corruption           1
year                 0
iso3_code            0
region               0
income_level         0
gdp_per_capita       5
unemployment_rate    5
dtype: int64

In [76]:
# Mostramos los registros de Happiness que no tienen
# datos de desempleo disponibles en el Banco Mundial
happiness_economic[
    happiness_economic["unemployment_rate"].isnull()
][[
    "country",
    "iso3_code",
    "year",
    "happiness_score",
    "unemployment_rate"
]]

,country,iso3_code,year,happiness_score,unemployment_rate
66,Kosovo,XKX,2015,5.589,NaN
229,Kosovo,XKX,2016,5.401,NaN
384,Kosovo,XKX,2017,5.279,NaN
525,Kosovo,XKX,2018,5.662,NaN
660,Kosovo,XKX,2019,6.100,NaN


De las 150 ausencias de desempleo que había en el dataset completo del Banco Mundial, al cruzarlo con Happiness solo nos afectan 5 filas, y son todas del mismo país

# Extracción del gasto sanitario desde la API del Banco Mundial

Como tercer indicador económico se analiza el gasto sanitario como porcentaje del PIB.

Este indicador permitirá estudiar si existe alguna relación entre la inversión en salud y los niveles de felicidad de los países.

Primero se realiza una prueba con España antes de extraer los datos para todos los países.